app/repositories/ingame_logins.py

```diff
@@ -0,0 +1,213 @@
+from __future__ import annotations
+
+import textwrap
+from typing import Any
+from typing import Optional
+
+import app.state.services
+
+# +--------------+------------------------+------+-----+---------+-------+
+# | Field        | Type                   | Null | Key | Default | Extra |
+# +--------------+------------------------+------+-----+---------+-------+
+# | id           | int                    | NO   | PRI | NULL    |       |
+# | userid       | int                    | NO   |     | NULL    |       |
+# | ip           | varchar(45)            | NO   |     | NULL    |       |
+# | osu_ver      | date                   | NO   |     | NULL    |       |
+# | osu_stream   | varchar(11)            | NO   |     | NULL    |       |
+# | datetime     | datetime               | NO   |     | NULL    |       |
+# +--------------+------------------------+------+-----+---------+-------+
+
+READ_PARAMS = textwrap.dedent(
+    """\
+        id, userid, ip, osu_ver, osu_stream, datetime
+    """,
+)
+
+
+async def create(userid: int, ip: str, osu_ver: str, osu_stream: str) -> dict[str, Any]:
+    """Create a new login entry in the database."""
+    query = f"""\
+        INSERT INTO ingame_logins (userid, ip, osu_ver, osu_stream, datetime)
+             VALUES (:userid, :ip, :osu_ver, :osu_stream, NOW())
+    """
+    params = {
+        "userid": userid,
+        "ip": ip,
+        "osu_ver": osu_ver,
+        "osu_stream": osu_stream,
+    }
+    rec_id = await app.state.services.database.execute(query, params)
+
+    query = f"""\
+        SELECT {READ_PARAMS}
+          FROM ingame_logins
+         WHERE id = :id
+    """
+    params = {
+        "id": rec_id,
+    }
+    rec = await app.state.services.database.fetch_one(query, params)
+    assert rec is not None
+    return dict(rec)
+
+
+async def fetch_one(
+    id: Optional[int] = None,
+    userid: Optional[int] = None,
+    ip: Optional[str] = None,
+    osu_ver: Optional[str] = None,
+    osu_stream: Optional[str] = None,
```

Comment:
only params which uniquely identify an ingame login should be accepted into fetch_one()

params which identify multiple should be used with fetch_many()

Reply:
(similarly, fetch_many should not have params which uniquely identify a single record - which seems correctly done in this pr)